pkg/netceptor/packetconn_test.go

```diff
@@ -0,0 +1,298 @@
+		{"Service Too Long Error", "123456789", "service name 123456789 too long", "Expected service name too long error, but got %v"},
+		{"Service Already Listening Error", "ping", "service ping is already listening", "Expected service ping is already listening, but got: %v"},
+	}
+
+	for _, testCase := range listenPacketTestCases {
+		ctx := context.Background()
+		netc := netceptor.New(ctx, "node")
+
+		t.Run(testCase.name, func(t *testing.T) {
+			_, err := netc.ListenPacket(testCase.service)
+			if testCase.expectedErr == "" && err != nil {
+				t.Errorf(testCase.failedTestString, err)
+			}
+			if testCase.expectedErr != "" && err != nil && err.Error() != testCase.expectedErr {
+				t.Errorf(testCase.failedTestString, err)
+			}
+			if testCase.expectedErr != "" && err == nil {
+				t.Errorf(testCase.failedTestString, err)
+			}
+		})
+	}
+}
+
+// TestListenPacketAndAdvertise test the ListenPacketAndAdvertise method.
+func TestListenPacketAndAdvertise(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	listenPacketTestCases := []struct {
+		name             string
+		service          string
+		tags             map[string]string
+		expectedErr      string
+		failedTestString string
+	}{
+		{"Success", "test", map[string]string{}, "", "Expected no error, but got: %v"},
+		{"Success empty service", "", map[string]string{}, "", "Expected no error, but got: %v"},
+		{"Service Too Long Error", "123456789", map[string]string{}, "service name 123456789 too long", "Expected service name too long error, but got %v"},
+		{"Service Already Listening Error", "ping", map[string]string{}, "service ping is already listening and advertising", "Expected service ping is already listening and advertising, but got: %v"},
+	}
+
+	for _, testCase := range listenPacketTestCases {
+		ctx := context.Background()
+		netc := netceptor.New(ctx, "node")
+
+		t.Run(testCase.name, func(t *testing.T) {
+			_, err := netc.ListenPacketAndAdvertise(testCase.service, testCase.tags)
+			if testCase.expectedErr == "" && err != nil {
+				t.Errorf(testCase.failedTestString, err)
+			}
+			if testCase.expectedErr != "" && err != nil && err.Error() != testCase.expectedErr {
+				t.Errorf(testCase.failedTestString, err)
+			}
+			if testCase.expectedErr != "" && err == nil {
+				t.Errorf(testCase.failedTestString, err)
+			}
+		})
+	}
+}
+
+// TestPacketConn tests both NewPacketConnWithConst and NewPacketConn methods.
+func TestPacketConn(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mockNetceptorForPacketConn := mock_netceptor.NewMockNetcForPacketConn(ctrl)
+
+	packetConnTestCases := []struct {
+		name                   string
+		service                string
+		expectedCall           func(ctx context.Context)
+		funcCall               func(pc netceptor.PacketConner) interface{}
+		expectedReturnVal      interface{}
+		unexpectedReturnValMsg string
+	}{
+		{
+			"GetLocalService Success",
+			"test",
+			func(ctx context.Context) {
```

Comment:
Ill leave this decision up to you, but, i wouldnt be against defining all these functions outside this test in order to keep this test case table smaller and easier to read. 

If you dont feel that will make a readability difference then im ok keeping it as is.

Reply:
happy to keep it the way it is